doc/neps/nep-0054-simd-cpp-highway.rst

```diff
@@ -0,0 +1,480 @@
+but not limited to:
+
+- Maintainability, domain expertise availability, ease of onboarding new
+  contributor, and other social aspects,
+- Key technical differences and constraints that may impact NumPy's internal
+  design or performance,
+- Build system related aspects and impact on our plans for upstreaming CPU and
+  compiler feature detection and multi-target library building into Meson,
+- Release timing related aspects.
+
+Out of scope (at least for now) is revisiting other aspects of our current SIMD
+support strategy:
+
+- accuracy vs. performance trade-offs when adding SIMD support to a function
+- use of SVML and x86-simd-sort (and possibly its equivalents for aarch64)
+- pulling in individual bits or algorithms of Highway (as in `gh-24018`_) or
+  SLEEF (as discussed in that same PR)
+
+
+Usage and Impact
+----------------
+
+N/A - there will be no significant user-visible changes.
+
+
+Backward compatibility
+----------------------
+
+There will be no changes in user-facing Python or C APIs: all the methods to
+control compilation and runtime CPU feature selection should remain, although
+there may be some changes due to moving to Meson and C++ without regards to the
+Highway/Universal Intrinsics choice.
+
+On Windows, MSVC may have to be avoided, as a result of Highway's use of
+pragmas which are less well supported by MSVC. This means that we likely have
+to build our wheels with clang-cl or Mingw-w64. Both of those should work - we
+merged clang-cl support a while back (see `gh-20866`_), and SciPy builds with
+Mingw-w64. It may however impact other redistributors or end users who build
+from source on Windows.
+
+The potential impact on licensing because of Highway's use of the Apache 2.0
+license needs discussing before Highway can be accepted. Or Highway would have
+to have to dual-license under Apache 2.0 and an MIT/BSD type license.
+
+
+High-level considerations
+-------------------------
+
+.. note::
+
+   Currently this section attempts to cover each topic separately, and
+   comparing the future use of a NumPy-specific C++ implementation vs. use of
+   Google Highway with our own numerical routines on top of that. It does not
+   (yet) assume a decision or proposed decision is made. Hence this NEP is not
+   "this is proposed" with another option in the Alternatives section, but
+   rather a side-by-side comparison.
+
+    
+Development effort and long-term maintainability
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Moving to Highway is likely to be a significant development effort.
+Longer-term, this will hopefully be offset by Highway itself having more
+maintainer bandwidth to deal with ongoing issues in compiler support and adding
+new platforms. It should increase the bus factor of SIMD support in NumPy to
+some extent. That said - the Highway repository also seems to show that its bus
+factor is low - it was 1 until the start of 2023, and now it's perhaps 2.
```

Comment:
FYI 12 Google engineers have open-sourced commits to Highway. We have two others besides myself able to approve and merge changes.

Reply:
Thanks for the context. I can add some more detail/stats. It's probably similar for NumPy; the bus factor calculation is based on commit history and probably still right though, that's more than number of people with a few commits and commit rights.

Reply:
Fair point. Currently there is no need for others to do what I am doing because it's my full-time role. If I were actually to be hit by a bus, it is likely that one or several of the others would take over :)

Reply:
This is where the line from Highway README page `This is not an officially supported Google product` gets a little concerning. I don't see this on other google open source projects though. Its very well possible I might be reading too much into it, but do you mind explaining what's different about Highway?

Reply:
That disclaimer is on most open source projects that have Googler contributors.

Reply:
C.f. [this search on Github](https://github.com/search?q=%22This+is+not+an+officially+supported+Google+product%22&type=code). It's very boilerplate.

Reply:
Its boilerplate for what google calls "pet" or "hobby" projects (see  https://opensource.google/documentation/reference/releasing#faq) which requires them to have this disclaimer. From: https://opensource.google/documentation/reference/releasing/publishing .

> We do this to help set appropriate expectations for users of these projects. Projects that include this label may not be staffed like larger, supported products such as Android, Chromium, or Go are. Support and/or new releases may be limited.

Highway isn't as big a project as Android or Chromium or Go, so probably doesn't matter.

Reply:
As briefly mentioned in the meeting: Highway is an engineering-led project that meets a developer need. The fact that it is not a user-facing product subject to business-level decisions about when to (dis)continue a project can be seen as helpful :)

Reply:
I suggest we remove the last two sentences from the NEP. We cannot predict future developments. so we should leave the more vague hope.